File: benchmark/benchmark.jl
#=
Julia array-summation benchmarks.

Refer to `README.md` for installation instructions for the requirments.
=#

cd(@__DIR__)

using AProgrammingLanguageForScientists, PyCall

@pyimport numpy as np

include("src.jl")

# Set benchmarking defaults to ensure 10,000 samples are always collected
BenchmarkTools.DEFAULT_PARAMETERS.samples = 10_000

results = Dict()

A = rand(1_000_000)

results["sum"] = ExecutionTime(A, sum)
results["naivesum"] = ExecutionTime(A, naivesum)
results["improvedsum"] = ExecutionTime(A, improvedsum)
results["numpy.sum"] = ExecutionTime(A, np.sum)

write("benchmarks_julia.json", JSON.json(results))

File: benchmark/benchmark.py
"""
Python array-summation benchmarks.

Refer to `README.md` for installation instructions for the requirments.
"""
import json
import numpy as np


def saveresult(f, res):
    results[f] = {"mean": res.average, "std": res.stdev}


results = {}

A = np.random.rand(1_000_000)

# numpy.sum
%timeit -o -n 1 -r 10_000 np.sum(A)
saveresult("numpy.sum", _)

A = list(A)

# sum
%timeit -o -n 1 -r 1_000 sum(A)
saveresult("sum", _)


# naivesumpy
def naivesum(A):
    accumulator = 0.

    for element in A:
        accumulator += element

    return accumulator


%timeit -o -n 1 -r 1_000 naivesum(A)
saveresult("naivesum", _)


with open("benchmarks_python.json", "w") as f:
    json.dump(results, f)
